### PR TITLE
VP9e - use C++ primitives (memcpy => std::copy)

### PR DESCRIPTION
**Pull request by ilapichev**
Issue: MDP-50976

### CODE REVIEW

File: _studio/mfx_lib/encode_hw/vp9/include/mfx_vp9_encode_hw_ddi.h [hunk 1]
```diff
@@ -185,22 +185,27 @@ typedef struct tagENCODE_CAPS_VP9
...
-    inline void AddSeqHeader(mfxU32 width,
+    inline mfxStatus AddSeqHeader(mfxU32 width,
         mfxU32 height,
         mfxU32 FrameRateN,
         mfxU32 FrameRateD,
         mfxU32 numFramesInFile,
         mfxU8* pBitstream,
         mfxU32 bufferSize)
     {
+    	if (bufferSize == 0)
```
Comment (mgonchar) [inline _studio/mfx_lib/encode_hw/vp9/include/mfx_vp9_encode_hw_ddi.h]: it should be
```
if (bufferSize < sizeof(ivf_file_header))
```

File: _studio/mfx_lib/encode_hw/vp9/src/mfx_vp9_encode_hw.cpp [hunk 2]
```diff
@@ -915,10 +915,16 @@ mfxStatus MFXVideoENCODEVP9_HW::GetVideoParam(mfxVideoParam *par)
     return MFX_ERR_NONE;
 }
 
-inline void UpdatePictureHeader(mfxU32 frameLen, mfxU32 frameNum, mfxU8* pPictureHeader, mfxU32 bufferSize)
+inline mfxStatus UpdatePictureHeader(mfxU32 frameLen, mfxU32 frameNum, mfxU8* pPictureHeader, mfxU32 bufferSize)
 {
+    if (bufferSize == 0)
```
Comment (mgonchar) [inline _studio/mfx_lib/encode_hw/vp9/src/mfx_vp9_encode_hw.cpp]: same about condition check

File: _studio/mfx_lib/encode_hw/vp9/src/mfx_vp9_encode_hw.cpp [hunk 3]
```diff
@@ -915,10 +915,16 @@ mfxStatus MFXVideoENCODEVP9_HW::GetVideoParam(mfxVideoParam *par)
...
+        return MFX_ERR_MORE_DATA;
+
     mfxU32 ivf_frame_header[3] = {frameLen, frameNum << 1, 0x00000000};
-    memcpy_s(pPictureHeader, bufferSize, ivf_frame_header, sizeof (ivf_frame_header));
+    //std::copy(std::begin(ivf_frame_header),std::end(ivf_frame_header), pPictureHeader);
+    std::fill_n(pPictureHeader, 12, 0);
```
Comment (mgonchar) [inline _studio/mfx_lib/encode_hw/vp9/src/mfx_vp9_encode_hw.cpp]: no, it is wrong, previous one was correct, you don't assign zeroes, these are different vals

File: _studio/mfx_lib/encode_hw/vp9/src/mfx_vp9_encode_hw.cpp [hunk 4]
```diff
@@ -903,7 +903,7 @@ mfxStatus MFXVideoENCODEVP9_HW::GetVideoParam(mfxVideoParam *par)
         mfxExtBuffer *pLocalBuf = GetExtendedBuffer(m_video.ExtParam, m_video.NumExtParam, pOutBuf->BufferId);
         if (pLocalBuf && (pOutBuf->BufferSz == pLocalBuf->BufferSz))
         {
-            memcpy_s(pOutBuf, pOutBuf->BufferSz, pLocalBuf, pLocalBuf->BufferSz);
+            std::copy(pLocalBuf, pLocalBuf + pLocalBuf->BufferSz, pOutBuf);
```
Comment (mgonchar) [inline _studio/mfx_lib/encode_hw/vp9/src/mfx_vp9_encode_hw.cpp]: you need to do pointer to bytes cast before using this pointer arithmetic

File: _studio/mfx_lib/encode_hw/vp9/src/mfx_vp9_encode_hw_ddi.cpp [hunk 5]
```diff
@@ -441,14 +441,15 @@ namespace MfxHwVP9Encode
...
-                    bufferSizeBytes);
-                ivfHeaderSize += IVF_SEQ_HEADER_SIZE_BYTES;
+            	mfxStatus sts = AddSeqHeader(task.m_frameParam.width,
+                                             task.m_frameParam.height,
+                                             par.mfx.FrameInfo.FrameRateExtN,
+                                             par.mfx.FrameInfo.FrameRateExtD,
+                                             0,
+                                             localBuf.pBuffer,
+                                             bufferSizeBytes);
+            	if (sts == MFX_ERR_NONE)
```
Comment (mgonchar) [inline _studio/mfx_lib/encode_hw/vp9/src/mfx_vp9_encode_hw_ddi.cpp]: no, if it returns error, I think we should stop the execution. i.e. use macro MFX_CHECK_STS (or just write manual code if don't have this macro in scope)

File: _studio/mfx_lib/encode_hw/vp9/include/mfx_vp9_encode_hw_ddi.h [hunk 6]
```diff
@@ -185,22 +185,27 @@ typedef struct tagENCODE_CAPS_VP9
...
+    		return MFX_ERR_MORE_DATA;
+
         mfxU32   ivf_file_header[8] = { 0x46494B44, 0x00200000, 0x30395056, width + (height << 16), FrameRateN, FrameRateD, numFramesInFile, 0x00000000 };
-        memcpy_s(pBitstream, bufferSize, ivf_file_header, sizeof (ivf_file_header));
+        std::copy(std::begin(ivf_file_header),std::end(ivf_file_header), pBitstream);
+
+        return MFX_ERR_NONE;
     };
 
     inline void AddPictureHeader(mfxU8* pBitstream, mfxU32 bufferSize)
```
Comment (dvrogozh) [inline _studio/mfx_lib/encode_hw/vp9/include/mfx_vp9_encode_hw_ddi.h]: this prototype should also be changed

**Pull request closed without merging**